network/src/compress.rs

```diff
@@ -0,0 +1,178 @@
+use bytes::{Bytes, BytesMut};
+use snap::{Decoder as SnapDecoder, Encoder as SnapEncoder};
+use tokio::codec::{Decoder, Encoder, LengthDelimitedCodec};
+
+use std::io;
+
+const SKIP_COMPRESS_SIZE: usize = 40 * 1024;
+
+/// Compressed decompression structure
+///
+/// If you want to support multiple compression formats in the future,
+/// you can simply think that 0b1000 is in snappy format and 0b0000 is in uncompressed format.
+///
+/// # Message in Bytes:
+///
+/// +---------------------------------------------------------------+
+/// | Bytes | Type | Function                                       |
+/// |-------+------+------------------------------------------------|
+/// |   0   |  u1  | Compress: true 1, false 0                      |
+/// |       |  u7  | Reserved                                       |
+/// +-------+------+------------------------------------------------+
+/// |  1~   |      | Payload (Serialized Data with Compress)        |
+/// +-------+------+------------------------------------------------+
+#[derive(Clone, Debug)]
+struct Message {
+    inner: BytesMut,
+}
+
+impl Message {
+    /// Init
+    fn init() -> Self {
+        Message {
+            inner: BytesMut::from(&[0u8][..]),
+        }
+    }
+
+    /// Compress message
+    fn compress(&mut self, input: Bytes) {
+        if input.len() > SKIP_COMPRESS_SIZE {
+            match SnapEncoder::new().compress_vec(&input) {
+                Ok(res) => {
+                    self.inner.unsplit(BytesMut::from(res));
+                    self.set_compress_flag(true);
+                }
+                Err(_) => {
```

Comment:
There is only one situation: 
> The total number of bytes to compress exceeds 2^32 - 1.

I don't think we will encounter this kind of problem, so I ignored the past. If there is not enough memory, it will panic